File: 알고리즘/그래프 이론/1926.cpp
// 그림

#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>
using namespace std;

class BFS{
private:
    int n, m;
    int** picture;
    int** visited;
    queue<pair<int, int>> q;
    vector<int> countList;
    int moveX[4] = {1, -1, 0, 0};
    int moveY[4] = {0, 0, 1, -1};

public:
    BFS(int n, int m){
        this->n = n; this->m = m;

        picture = new int*[n];
        visited = new int*[n];

        for(int i = 0; i < n; i++){
            picture[i] = new int[m];
            visited[i] = new int[m];
        }
    }
    void drawPicture(){
        for(int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                cin >> picture[i][j];
                visited[i][j] = 0;
            }
        }
    }
    void searchPicture(){
        for(int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                if(picture[i][j] && !visited[i][j])
                    startBFS(i, j);
            }
        }
    }
    void startBFS(int s, int u){
        visited[s][u] = 1;
        q.push(make_pair(s, u));

        int count = 0;

        while(!q.empty()){
            count++;

            int x = q.front().first;
            int y = q.front().second;
            q.pop();

            for(int i = 0; i < 4; i++){
                int newX = x + moveX[i];
                int newY = y + moveY[i];

                if(newX >= 0 && newY >= 0 && newX < n && newY < m && picture[newX][newY] && !visited[newX][newY]){
                    visited[newX][newY] = 1;
                    q.push(make_pair(newX, newY));
                }
            }
        }
        countList.push_back(count);
    }
    void printCount(){
        sort(countList.begin(), countList.end());

        cout << countList.size() << "\n";

        if(!countList.size()){
            cout << 0 << endl;
        }
        else {
            cout << countList.back() << endl;
        }
    }
};

int main(){
    int n, m;
    cin >> n >> m;
    
    BFS bfs(n, m);
    bfs.drawPicture();
    bfs.searchPicture();
    bfs.printCount();
}

File: 알고리즘/그래프 이론/1926.py
# 그림

import queue

class BFS:
    def __init__(self, n, m):
        self.__n = n
        self.__m = m
        self.__picture = []
        self.__countList = []
        self.__visited = [[False for i in range(m)] for i in range(n)]
        self.__q = queue.Queue()
        self.__moveX = [1, -1, 0, 0]
        self.__moveY = [0, 0, 1, -1]
    
    def drawPicture(self):
        for _ in range(self.__n):
            self.__picture.append(list(map(int, input().split(' '))))
    
    def searchPicture(self):
        for i in range(self.__n):
            for j in range(self.__m):
                if self.__picture[i][j] and not self.__visited[i][j]:
                    self.bfs(i, j)

    def bfs(self, s, u):
        self.__visited[s][u] = True
        self.__q.put([s, u])
        count = 0

        while not self.__q.empty():
            count += 1

            front = self.__q.get()
            x = front[0]
            y = front[1]

            for i in range(4):
                newX = x + self.__moveX[i]
                newY = y + self.__moveY[i]

                if 0 <= newX < n and 0 <= newY < m and self.__picture[newX][newY] and not self.__visited[newX][newY]:
                    self.__visited[newX][newY] = True
                    self.__q.put([newX, newY])

        self.__countList.append(count)

    def printCount(self):
        print(len(self.__countList))
        if len(self.__countList) == 0:
            print(0)
        else:
            print(max(self.__countList))

n, m = map(int, input().split(' '))

pic = BFS(n, m)
pic.drawPicture()
pic.searchPicture()
pic.printCount()